### PR TITLE
MINIFICPP-1439 - Startup without agent class, handle class update

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/http-curl/tests/C2RequestClassTest.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        return false;
+      }
+      ++it;
+    }
+    return true;
+  }
+
+ private:
+  std::mutex mtx_;
+  std::vector<utils::optional<std::string>> classes_;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/http-curl/tests/C2RequestClassTest.cpp:94]: I would like to distinguish between a heartbeat sending an empty agent class, and a heartbeat sending no agent class

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/http-curl/tests/C2RequestClassTest.cpp:94]: but I don't have strong feelings for this version, should I change it?

[the remark this answers is not in the document]
Comment (arpadboda) [reply extensions/http-curl/tests/C2RequestClassTest.cpp:94]: I think that's correct, but usecase-wise I see no difference between the two. 
Do you think there is any?

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/http-curl/tests/C2RequestClassTest.cpp:94]: well, I am leaning towards considering a heartbeat with `""` agent class an error (although we do not check for such cases), it all depends on what the constraints of the c2 protocol are, do you think an `""` agent class should have the same semantics as a "missing" agent class, or we could even abandon the "missing" agent class, and use the empty string to denote an agent with non-specified class?

File: libminifi/include/c2/C2Payload.h [hunk 2]
```diff
@@ -182,6 +197,8 @@ class C2Payload : public state::Update {
   Operation op_;
   bool raw_{ false };
   std::vector<char> raw_data_;
+  // TODO(adebreceni): isResponse is currently unused and incorrectly
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/c2/C2Payload.h]: done in #971

File: extensions/http-curl/tests/C2RequestClassTest.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    if (root.HasMember("operationId")) {
+      std::lock_guard<std::mutex> guard(mtx_);
+      acknowledged_operations_.insert(root["operationId"].GetString());
+    }
+    mg_printf(conn, "HTTP/1.1 200 OK\r\nContent-Type: "
+                    "text/plain\r\nContent-Length: 0\r\nConnection: close\r\n\r\n");
+    return true;
+  }
+
+  bool isAcknowledged(const std::string& operation_id) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/http-curl/tests/C2RequestClassTest.cpp]: done

**Review by arpadboda — APPROVED** [review]

**Pull request closed without merging**